System/Console/Haskeline/Backend/Win32/Echo.hs

```diff
@@ -0,0 +1,172 @@
+--            ('liftIO' . 'hSetInputEchoState' input)
+--            (const action)
+-- @
+hBracketInputEcho :: MonadException m => Handle -> m a -> m a
+hBracketInputEcho input action =
+  bracket (liftIO $ hGetInputEchoState input)
+          (liftIO . hSetInputEchoState input)
+          (const action)
+
+-- | Perform a computation with the handle's input echoing disabled. Before
+-- running the computation, the handle's input 'EchoState' is saved, and the
+-- saved 'EchoState' is restored after the computation finishes.
+--
+-- @
+-- hWithoutInputEcho input action =
+--   'hBracketInputEcho' input
+--                       ('liftIO' ('hSetInputEchoState' input ('hEchoOff' input)) >> action)
+-- @
+hWithoutInputEcho :: MonadException m => Handle -> m a -> m a
+hWithoutInputEcho input action =
+  hBracketInputEcho input
+                    (liftIO (hSetInputEchoState input (hEchoOff input)) >> action)
+
+-- | Create an @stty@ process, wait for it to complete, and return its output.
+hSttyRaw :: Handle -> String -> IO STTYSettings
+hSttyRaw input arg = do
+  let stty = (shell $ "stty " ++ arg) {
+        std_in  = UseHandle input
+      , std_out = CreatePipe
+      }
+  (_, mbStdout, _, rStty) <- createProcess stty
+  exStty <- waitForProcess rStty
+  case exStty of
+    e@ExitFailure{} -> throw e
+    ExitSuccess     -> maybe (return "") hGetContents mbStdout
+
+-- | A representation of the handle input's current echoing state.
+-- See, for instance, 'hEchoOff'.
+data EchoState
+  = MinTTY STTYSettings
+    -- ^ The argument to (or value returned from) an invocation of the @stty@
+    -- command-line utility. Most POSIX-like shells have @stty@, including
+    -- MinTTY on Windows. Since neither 'hGetEcho' nor 'hSetEcho' work on
+    -- MinTTY, when 'getInputEchoState' runs on MinTTY, it returns a value
+    -- built with this constructor.
+    --
+    -- However, native Windows consoles like @cmd.exe@ or PowerShell do not
+    -- have @stty@, so if you construct an 'EchoState' with this constructor
+    -- manually, take care not to use it with a native Windows console.
+  | DefaultTTY Bool
+    -- ^ A simple on ('True') or off ('False') toggle. This is returned by
+    -- 'hGetEcho' and given as an argument to 'hSetEcho', which work on most
+    -- consoles, with the notable exception of MinTTY on Windows. If you
+    -- construct an 'EchoState' with this constructor manually, take care not
+    -- to use it with MinTTY.
+  deriving (Eq, Ord, Show)
+
+-- | Indicates that the handle's input echoing is (or should be) off.
+hEchoOff :: Handle -> EchoState
+hEchoOff input = if minTTY input then MinTTY "-echo" else DefaultTTY False
+
+-- | Settings used to configure the @stty@ command-line utility.
+type STTYSettings = String
+
+-- | Is the current process attached to a MinTTY console (e.g., Cygwin or MSYS)?
+minTTY :: Handle -> Bool
+#if MIN_VERSION_Win32(2,5,0)
+minTTY input = unsafePerformIO $ withHandleToHANDLE input isMinTTYHandle
```

Comment:
Can we make this `Handle -> IO Bool`, and remove the `unsafePerformIO` and `NOINLINE`?